netket/experimental/hilbert/fermions_2nd.py

```diff
@@ -0,0 +1,112 @@
+from re import L
+from typing import Optional, List, Union
+from xml.sax.handler import property_declaration_handler
+from netket.hilbert.fock import Fock
+from collections.abc import Iterable
+from netket.hilbert.homogeneous import HomogeneousHilbert
+import jax.numpy as jnp
+from netket.hilbert.tensor_hilbert import TensorHilbert
+import numpy as np
+from fractions import Fraction
+
+
+class SpinOrbitalFermions(HomogeneousHilbert):
+    def __init__(
+        self, n_orbitals: int, s: float = 0.0, n_fermions: Union[int, List[int]] = None
+    ):
+        r"""
+        Class of fermions in 2nd quantization with n_orbitals and spin s.
+        Samples of this hilbert space represent occupation numbers (0,1) of the orbitals.
+        Number of fermions can be fixed by setting n_fermions.
+        If the spin is different from 0, n_fermions can also be a list to fix the number of fermions per spin component.
+        Using this class, one can generate a tensor product of fermionic hilbert spaces that distinguish particles with different spin.
+
+        Args:
+            n_orbitals (required): number of orbitals we store occupation numbers for. If the number of fermions per spin is conserved, the different spin configurations are not counted as orbitals and are handled differently.
+            s (float): spin of the fermions.
+            n_fermions (optional, int or list(int)): fixed number of fermions per spin (conserved). In the case n_fermions is an int, the total number of fermions is fixed, while for lists, the number of fermions per spin component is fixed.
+
+        Returns:
+            A SpinOrbitalFermions object
+        """
+        spin_size = round(2 * s + 1)
+        spin_states = np.empty(spin_size)
+        for i in range(spin_size):
+            spin_states[i] = 2 * i - round(2 * s)
+        spin_states = spin_states.tolist()
+
+        total_size = n_orbitals * spin_size
+        if n_fermions is None:
+            hilbert = Fock(n_max=1, N=total_size)
+        else:
+            if isinstance(n_fermions, int):
+                hilbert = Fock(n_max=1, N=total_size, n_particles=n_fermions)
+            else:
+                if not isinstance(n_fermions, Iterable):
+                    raise ValueError("n_fermions must be iterable or int")
+                if len(n_fermions) != spin_size:
+                    raise ValueError(
+                        "list of number of fermions must equal number of spin components"
+                    )
+                spin_hilberts = [
+                    Fock(n_max=1, N=n_orbitals, n_particles=Nf) for Nf in n_fermions
+                ]
+                hilbert = TensorHilbert(*spin_hilberts)
+
+        # internally, we store a Fock space or a TensorHilbert of Fock spaces
+        self._fock = hilbert
```

Comment:
You can put a docstring to internal fields. Not changes much, but gives a nice docstring in the auto-suggestion of vscode so why not.

```suggestion
        self._fock = hilbert
        """Internal representation of this Hilbert space, which is either a Fock or TensorHilbert."""
```